### PR TITLE
sql: assert no dependencies on storage or CGo

### LINKED ISSUE #30001: *: avoid sql -> c-dep dependency edges
Some of cockroach's code is likely useful outside of just the cockroach binary. Most obvious is some of our general purpose util packages, but additionally anyone building tools that interact with Cockroach or its data such as api responses, on-disk data, backups, etc may want to reuse our defined protobufs and structs as well as the helpers and other functions we've defined around them. Such external tools, both from first or third parties, are common in healthy, thriving OSS ecosystems, and we ideally should, where practical, make it as easy as possible to use our reusable helpers both in and outside the cockroach repo.

However a point of friction currently is that Cockroach uses `make` to build, including to run just-in-time code generation and to handle building c dependencies for linking by cgo. Non-make driven builds of the majority of the codebase simply do not work. Solving the generation problem has been previously addressed by the creation of the `cockroach-gen` mirror of the repo, maintained with pre-generated code by CI.

However cgo remains a problem, rendering any package with transitive dependencies on cgo unusable outside the cockroach binary. This is unlikely to change given the platform-specific nature of our c builds and how dependent they are on `make`. However we can minimize the number of our Go packages that become impossible to consume externally by avoiding depending, including transitively, on cgo.  

Current examples:
`mvcc` encoding in `storage/engine` is used in sql and elsewhere, but since `storage/engine` also handles interfacing with rocksdb, pulling in pure-Go mvcc logic pulls in cgo.
`distsqlrun` uses interfaces that are defined in `engine` and even creates stores that are backed by rocksdb, thus depending on in rocksdb directly.

This will likely require care in splitting some of the low-level packages, defining interfaces and splitting cgo-backed implementations out into separate packages. In the of the temp-store example, it would also require some form of dependency injection as it is actually calling directly into a cgo-backed implementation.

Additionally all of this is bound to rot if not enforced by machines, as large dependency graphs can easily hide edges from human inspection, so we'll want to make a list of packages that a lint inspects, to ensure their transitive dependency closures remain free of cgo imports.

### PR DESCRIPTION
**Pull request by jordanlewis**
No longer necessary for sql to depend on either storage or CGo - let's keep it that way!

All commits but last one from dependent PRs.

Closes #30001.

### CODE REVIEW

**Comment by jordanlewis** [comment]
Yay, this can merge now.

**Review by dt — APPROVED** [review]
🎉

**Comment by dt** [comment]
@lopezator This change broke the dependency from `sql` to `storage` and thus any transitive edge to our C++ dependencies that use a `make`-driven build process, like rocksdb, and banned them from coming back into the dependencies of `sql`, but that doesn't include plain old (go-built) cgo.

I think @mjibson has been looking at making the `parser` package cgo-free but it hasn't proved straightforward: anything that pulls in stdlib's `tls`, like our `base` config package or the `lib/pq` library that our errors package uses pick up a cgo dep via the stdlib's `crypto/x509`. Also, anything that pulls in our `build` info package, including our `log` package, will see cgo as well, since there's a little inline C in there to identify which cgo compiler is in use. 

We _might_ be able to eliminate that `build` C import at this point -- now that we're committed to `make` driven builds, we can likely just have `make` inject the c++ compiler used instead of compiling a little inline program to figure it out (cc @benesch?).

However we still have cgo deps via various stdlib packages, namely via `tls` and `crypto/x509` and via `net`. We can cut the `tls` edge, both from the `pgerror` to `lib/pq` path, since the pgerror helper is only used in tests, and then we could move the `base.DocsURL` helper, but eventually we'll hit stdlib `net` and whatever other paths I haven't seen. I'm not sure we can practically avoid any stdlib packages in all our transitive deps, so if they use cgo, we might be out of luck? I dunno, @mjibson has been looking at this longer than I have so maybe he has further insight.

**Comment by benesch** [comment]
> We might be able to eliminate that build C import at this point -- now that we're committed to make driven builds, we can likely just have make inject the c++ compiler used instead of compiling a little inline program to figure it out (cc @benesch?).

Definitely an option, but I'm not sure it buys us much. These packages whose dependency sets we're minimizing shouldn't be depending on `build` anyway. My laptop is currently out of commission while it upgrades to Mojave, but I'll do some digging on where the edge to `build` is coming from.

It's ok to use the net package. Go has some crazy shenanigans where it will bundle its own DNS resolver that doesn't require cgo if you compile net with `CGO_ENABLED=0`.

tl;dr I think making it so that these library packages can be compiled without cgo is totally within reach.

**Comment by dt** [comment]
the edge to build is pervasive: the most direct edge here is via the `base` package but everything gets to logging eventually and logging get to build to determine if it is a release/dev build. I guess we could buildtag out the inline cgo in `build` ourselves for non-cgo builds?

**Comment by lopezator** [comment]
Stdlib (CGO or not, as @benesch shouldn't be a problem, I think).

The offending parts I detected for my particular usecase of **sql** package to use in:

https://github.com/lopezator/sqlfmt

Are:

https://github.com/cockroachdb/cockroach/blob/master/pkg/util/tracing/annotate.go#L17-L19

And the one mentioned by @dt :

https://github.com/cockroachdb/cockroach/blob/master/pkg/build/info.go#L26-L35

Removing them from makes my sqlfmt free of any CGO, C go, and thus, capable of building using standard go and without needing any C paraphernalia, i.e:

`GOOS=windows GOARCH=amd64 go build cmd/sqlfmt/main.go`
`GOOS=darwin GOARCH=amd64 go build cmd/sqlfmt/main.go`

**Comment by mjibson** [comment]
@benesch the build package is annoyingly common. Over 100 packages have transitive dependencies on it including almost everything in sql, storage, util, server. It's because of two reasons: the log package uses it on crashes and a few things in SQL use it to get the docs URL of the current version. I tried https://github.com/mjibson/cockroach/commit/d1a209d36e0898b4fe5bfd457935491f38abb300 at one point which causes build to poke into other packages to remove the dependency on it, but I'm not convinced it's a good idea, or maybe there's a better way.

**Comment by dt** [comment]
@mjibson https://github.com/cockroachdb/cockroach/pull/31467
